README.en

```diff
@@ -0,0 +1,357 @@
+aimed at quick checking a huge amount of text against regular expressions.
+The library checks whether the text matches the search pattern with the 
+speed 400 MB/s regardless of regular expression complexity on our hardware.
+Several search patterns can be united for text checking, it allows to perform
+text checking against about ten patterns at a time at the same speed.
+
+Since Pire examines each character only once, without any lookaheads
+or rollbacks, spending about five machine instructions per each character,
+it can be used even in realtime tasks.
+
+On the other hand, Pire has limited functionality compared to
+other regexp libraries. Pire does not support Perl conditional
+regexps, lookaheads & backtrackings, greedy/non-greedy matches; neither
+all capturing facilities (limited version of capturing can be implemented
+as described below).
+
+Full details of methods that allow Pire to be so fast are described in
+Dragon Book. We recommend to read this book for a better principle understanding.
+
+
+Pire was developed in Yandex (http://company.yandex.ru/) as a part of its
+web crawler and became an independent project. 
+
+
+Quick start
+=============
+
+#include <stdio.h>
+#include <vector>
+#include <pire/pire.h>
+
+Pire::Scanner CompileRegexp(const char* pattern)
+{
+	// Convert the template to UCS4
+	std::vector<Pire::wchar32> ucs4;
+	Pire::Encodings::Utf8().FromLocal(pattern, pattern + strlen(pattern), std::back_inserter(ucs4));
+	// или другая кодировка
+
+	return Pire::Lexer(ucs4.begin(), ucs4.end())
+		.AddFeature(Pire::Features::CaseInsensitive())	// to ignore case
+		.SetEncoding(Pire::Encodings::Utf8())		// set the input string encoding
+		.Parse() 					// parse the regular expression
+		.Surround()					// to disable the PCRE_ANCHORED logic
+		.Compile<Pire::Scanner>();			// compile the regular expression
+}
+
+bool Matches(const Pire::Scanner& scanner, const char* ptr, size_t len)
+{
+	return Pire::Runner(scanner)
+		.Begin()	// the start of the string
+		.Run(ptr, len)	// the string
+		.End();		// the end of the string
+		// implicitly cast to bool
+}
+
+int main()
+{
+	char re[] = "hello\\s+w.+d$";
+	char str[] = "Hello world";
+
+	Pire::Scanner sc = CompileRegexp(re);
+
+	bool res = Matches(sc, str, strlen(str));
+
+	printf("String \"%s\" %s \"%s\"\n", str, (res ? "matches" : "doesn't match"), re);
+		
+	return 0;
+}
+
+
+Running `MATCH`
+==================
+
+The compiled regular expression is stored in the scanner which is the state machine
+in the form optimized for moving across the strings. There are several scanners with
+approximately the same operation in Pire (for example, Scanner, SlowScanner и SimpleScanner).
+
+Each scanner has the State type that stores the state, and methods Initialize(State&)
+to return the initial machine state, Next(State&, char) to take the machine to the next state,
+Final(const State&) to report whether the state is valid. To check whether the string matches
```

Comment:
It is better to use the word 'final' instead of 'valid':

> Final(const State&) to report whether the state is final (ie. the processed text is accepted by the machine).